File: Digital_Joper_mala/app/src/main/java/com/mala/digital_joper_mala/Database/Database_helper.java
package com.mala.digital_joper_mala.Database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Database_helper extends SQLiteOpenHelper {

    public static final String DATABASE_NAME = "mantra.db";
    public static final int VERSION = 2;
    public static final String TABLE_NAME = "mantra_table";

    public Database_helper(@Nullable Context context) {
        super(context, DATABASE_NAME, null, VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {

        db.execSQL("CREATE TABLE " +TABLE_NAME+ "(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, mantra TEXT)");

    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        db.execSQL("DROP TABLE IF EXISTS " +TABLE_NAME);
        onCreate(db);

    }

    public void insert(String title, String mantra){

        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues values = new ContentValues();

        values.put("title", title);
        values.put("mantra", mantra);

        db.insert(TABLE_NAME,null, values);

        db.close();

    }

    public boolean delete(String title){

        SQLiteDatabase db = this.getWritableDatabase();

        int result =   db.delete(TABLE_NAME,"title = ?", new String[]{title});
        db.close();

        return result > 0;
    }

    public List<HashMap<String, String>> getData(){

        List<HashMap<String, String>> list = new ArrayList<>();
        SQLiteDatabase db = this.getReadableDatabase();

        Cursor cursor = db.rawQuery("SELECT title,mantra FROM "+TABLE_NAME, null);

        while (cursor.moveToNext()){

            HashMap<String, String> map = new HashMap<>();

            map.put("title", cursor.getString(cursor.getColumnIndexOrThrow("title")));
            map.put("mantra", cursor.getString(cursor.getColumnIndexOrThrow("mantra")));
            list.add(map);

        }

        cursor.close();
        db.close();

        return list;
    }
}

File: Digital_Joper_mala/app/src/main/java/com/mala/digital_joper_mala/Helper/DatabaseHelper.kt
package com.mala.digital_joper_mala.Helper

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

abstract class DatabaseHelper<T>(
    val context : Context,
    val dbName : String,
    val dbVersion : Int
) : SQLiteOpenHelper(context, dbName, null, dbVersion
) {
    
    abstract fun onInsert() : Long
    abstract fun onGetAll(page : Int) : List<T>
    abstract fun onGetOne() : T
    abstract fun onDelete() : Boolean
    abstract fun onDeleteAll() : Boolean

    /* common tools */

    protected fun createTable(
        tableName: String,
        columnName : String
    ) : String{

        val sql = """
            CREATE TABLE IF NOT EXISTS $tableName 
            (id INTEGER PRIMARY KEY AUTOINCREMENT,
            $columnName
            )
        """.trimIndent()

        return sql
    }

    protected fun upgradeTable(
        tableName: String
    ) : String{

        return "DROP TABLE IF EXISTS $tableName"

    }

    protected fun insertData(
        tableName : String,
        values : ContentValues
    ) : Long{

       return writableDatabase.insert(tableName, null, values)

    }

    protected fun getData(
        tableName: String,
        otherSQL : String,
        selectionArgs : Array<String>?
    ) : Cursor{

        val result = if (otherSQL.isEmpty()){

            readableDatabase.rawQuery("SELECT * FROM $tableName", selectionArgs)

        }else{

            readableDatabase.rawQuery("SELECT * FROM $tableName $otherSQL", selectionArgs)

        }

        return result

    }

    protected fun deleteOne(
        tableName: String,
        whereClause : String,
        whereArgs : Array<String>?
    ) : Boolean{

        return writableDatabase.delete(tableName, whereClause, whereArgs) > 0

    }

    protected fun deleteAll(
        tableName: String
    ) : Boolean{

        return writableDatabase.delete(tableName, null, null) > 0

    }

    protected fun updateData(
        tableName: String,
        values: ContentValues,
        whereClause: String?,
        whereArgs : Array<String>?
    ) : Boolean{

        return writableDatabase.update(
            tableName,
            values,
            whereClause,
            whereArgs
        ) > 0

    }

    protected fun isExits(
        tableName: String,
        columName : String,
        whereClause: String,
        whereArgs: Array<String>?,

    ) : Boolean{

        val sql = readableDatabase.query(
            tableName,
            arrayOf(columName),
            whereClause,
            whereArgs,
            null,
            null,
            null,
            "1"
        )

        return sql.use { it.moveToFirst() }

    }

}